### ISSUE TITLE
Desired GMRQ behavior when scoring discrete trajectories?

### ISSUE THREAD

**Issue by maxentile** [msmbuilder/msmbuilder#787]
Hi!

If I ask a `MarkovStateModel` to score discrete trajectories that never enter its largest strongly connected component, MSMBuilder raises an error instead of returning `0`.

```python
import numpy as np
import numpy.random as npr
from msmbuilder.msm import MarkovStateModel

# a lot of trajectories exploring one ergodic subspace
dtrajs_a = [npr.randint(low=0,high=50,size=10000) for i in range(100)]

# a few trajectories exploring a different ergodic subspace
dtrajs_b = [npr.randint(low=50,high=100,size=10000) for i in range(10)]

# fit MSM to all trajectories
dtrajs = dtrajs_a + dtrajs_b
msm = MarkovStateModel()
msm.fit(dtrajs)

# score on trajectories that never entered the largest strongly connected component
msm.score(dtrajs_b)
```

Results in:
```

MSM contains 2 strongly connected components above weight=1.00. Component 0 selected, with population 90.909091%
MSM contains 1 strongly connected component above weight=1.00. Component 0 selected, with population 100.000000%
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-5-aedb3cc56530> in <module>()
      1 # score on trajectories that never entered the largest strongly connected component
----> 2 msm.score(dtrajs_b)

.../msmbuilder/msm/msm.pyc in score(self, sequences, y)
    456 
    457         if self.mapping_ != m2.mapping_:
--> 458             V = self._map_eigenvectors(V, m2.mapping_)
    459             # we need to map this model's eigenvectors
    460             # into the m2 space

.../msmbuilder/msm/msm.pyc in _map_eigenvectors(self, V, other_mapping)
    475         self_inverse_mapping = {v: k for k, v in self.mapping_.items()}
    476         transform_mapping = _dict_compose(self_inverse_mapping, other_mapping)
--> 477         source_indices, dest_indices = zip(*transform_mapping.items())
    478 
    479         #print(source_indices, dest_indices)

ValueError: need more than 0 values to unpack
```

Should this return `0` instead?

Josh

**Comment by rmcgibbo** [comment]
yes, when the two spaces are disjoint the gmrq is not really well defined.

On Thu, Mar 31, 2016 at 9:52 AM, Josh Fass <notifications@github.com> wrote:

> Hi!
>
> If I ask a MarkovStateModel to score discrete trajectories that never
> enter its largest strongly connected component, MSMBuilder raises an error
> instead of returning 0.
>
> import numpy as npimport numpy.random as nprfrom msmbuilder.msm import MarkovStateModel
> # a lot of trajectories exploring one ergodic subspace
> dtrajs_a = [npr.randint(low=0,high=50,size=10000) for i in range(100)]
> # a few trajectories exploring a different ergodic subspace
> dtrajs_b = [npr.randint(low=50,high=100,size=10000) for i in range(10)]
> # fit MSM to all trajectories
> dtrajs = dtrajs_a + dtrajs_b
> msm = MarkovStateModel()
> msm.fit(dtrajs)
> # score on trajectories that never entered the largest strongly connected component
> msm.score(dtrajs_b)
>
> Results in:
>
>
> MSM contains 2 strongly connected components above weight=1.00. Component 0 selected, with population 90.909091%
> MSM contains 1 strongly connected component above weight=1.00. Component 0 selected, with population 100.000000%
> ---------------------------------------------------------------------------
> ValueError                                Traceback (most recent call last)
> <ipython-input-5-aedb3cc56530> in <module>()
>       1 # score on trajectories that never entered the largest strongly connected component
> ----> 2 msm.score(dtrajs_b)
>
> .../msmbuilder/msm/msm.pyc in score(self, sequences, y)
>     456
>     457         if self.mapping_ != m2.mapping_:
> --> 458             V = self._map_eigenvectors(V, m2.mapping_)
>     459             # we need to map this model's eigenvectors
>     460             # into the m2 space
>
> .../msmbuilder/msm/msm.pyc in _map_eigenvectors(self, V, other_mapping)
>     475         self_inverse_mapping = {v: k for k, v in self.mapping_.items()}
>     476         transform_mapping = _dict_compose(self_inverse_mapping, other_mapping)
> --> 477         source_indices, dest_indices = zip(*transform_mapping.items())
>     478
>     479         #print(source_indices, dest_indices)
>
> ValueError: need more than 0 values to unpack
>
> Should this return 0 instead?
>
> Josh
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/msmbuilder/msmbuilder/issues/787>
>



-- 
-Robert

**Comment by mpharrigan** [comment]
How about `np.nan`

**Comment by maxentile** [comment]
Ahh, thank you for clarifying! Makes sense that it would be undefined instead of zero here.

Since this situation can easily arise in cross-validation, some additional guidance for model selection could be helpful.

Example: diffusion in a two-well potential

![image](https://cloud.githubusercontent.com/assets/5759036/14282638/4b0be146-fb0d-11e5-8f3e-6699d63a5cf5.png)

We'd like to select between two candidate discretizations (aka a binary hyperparameter):

![image](https://cloud.githubusercontent.com/assets/5759036/14282649/4f596246-fb0d-11e5-9998-652618d8c2c9.png)

vs.

![image](https://cloud.githubusercontent.com/assets/5759036/14282653/544abe76-fb0d-11e5-83e2-189015e8b335.png)

(both have 50 bins, Matplotlib defaults just make the x-axis look bigger)

If we do 2-fold CV, and happen to split the trajectories from the two wells, we'll probably choose `discretization_fast`:
- Mean train/test scores for `discretization_fast`: `6.603`, `6.531`
- Mean train/test scores for `discretization_slow`:
`6.620`, `nan`

This is about the same result as leave-one-out:
- Mean train/test scores for `discretization_fast`: `6.637`, `6.301`
- Mean train/test scores for `discretization_slow`:
`6.620`, `nan`

The model that finely discretizes the actual slow coordinate has undefined "test" scores, since the rare transition process hasn't been sampled yet, and all trajectories stay in either one well or the other.

A related issue: the GMRQ-selected `MarkovStateModel` here reports that its largest strongly connected component contains 100% of the observed counts. This might encourage the mistaken interpretation that the simulations sampled transitions between metastable states (or that the model represents rates between them) -- esp. if we know that our starting configurations are drawn from different metastable states, but we can't directly examine the cluster boundaries. I would think we want to select the other model in this situation, which only retains clusters corresponding to one well, and "trims" clusters from the other well.

Is there additional guidance on how to do model selection in this situation?

([notebook](https://github.com/maxentile/msm-learn/blob/master/notebooks/CV%20issue.ipynb))

Thanks!

Josh

**Comment by mpharrigan** [comment]
Josh,
Thanks for the great example! Would you mind if we included those figures in our docs in the (currently empty) discussion of gmrq?

One comment. This doesn't entirely address the problem, but we've given some amount of thought into the type of cross validation we use. I think initially @rmcgibbo did k-fold cross validation and a reviewer presciently asked "well why'd you pick that?". In most contexts, "leave-one-out" cross validation is the gold standard and k-fold is an approximation. Here, that doesn't make sense. In true leave-one-out, you'd use all your counts to train a model, and your test set would be one count. The msm fit on the test data would not be diagonalized all that well by the train eigenvectors.

We generally do "shuffle split" cross validation now. You take your trajectories, and split it 50:50 test:train and iterate over the different permutations of such splits. 

This would "help" in your case by showing that trimming is only keeping 50% of the data. I put help in quotes because the score might still be nan and/or misleading.

You could try adding a pseudo count (in msm constructor set `prior_counts=1e-5, ergodic_cutoff='off'`). The actual value of the gmrq would depend on which you pick for the value of prior_counts, but at least it would suggest the correct state space decomposition

**Comment by jchodera** [comment]
> In true leave-one-out, you'd use all your counts to train a model, and your test set would be one count

I'd be a bit careful here, since this isn't entirely correct. Cross-validating on *counts* destroys the correlation structure in the data, so would be totally incorrect for a case in which you have strong correlation within each trajectory.  Cross-validating or bootstrapping by trajectories---or sets of trajectories---is one way to preserve the correlation structure in the data.

**Comment by jchodera** [comment]
For bootstrap or cross-validation sets, what about simply omitting synthetic datasets that don't sample the same state space?  This will obviously give you an underestimate that might be severe if you have a small threshold for ergodic trimming, but it might still be *useful*.

**Comment by mpharrigan** [comment]
> I'd be a bit careful here, since this isn't entirely correct. Cross-validating on counts destroys the correlation structure in the data, so would be totally incorrect for a case in which you have strong correlation within each trajectory

This is sortof the point I'm making. Consider having 10k fah-style (short) trajectories and doing leave-one-out (on the trajectory level)

**Comment by mpharrigan** [comment]
> For bootstrap or cross-validation sets, what about simply omitting synthetic datasets that don't sample the same state space? This will obviously give you an underestimate that might be severe if you have a small threshold for ergodic trimming, but it might still be useful.

Isn't this essentially what @maxentile 's code above ends up doing? If you ignore the data that samples e.g. a second well, gmrq suggests the wrong discretization
